### DISCUSSION TITLE
'flux registers' for edge-centered EMFs?

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2511] [Q&A] [answered]
Hi,

We are planning to implement face-centered MHD in Quokka. Is there a 'flux register' to do the correction at coarse-fine interfaces of edge-centered EMFs in AMReX? In Chombo, this can be accomplished with [`LevelFluxRegisterEdge`](http://colella.lbl.gov/Manuals/CHOMBO-RELEASE-3.3/classLevelFluxRegisterEdge.html), which has a similar API to `YAFluxRegister`.

Example:
<img width="372" alt="Screen Shot 2021-12-03 at 2 14 21 PM" src="https://user-images.githubusercontent.com/1781919/144539019-11d98c91-9ac2-4ed6-a347-96a6a0b97b0a.png">

There are a number of corner cases (see panels a-d copied below). (In [Athena++](https://arxiv.org/pdf/2005.06651.pdf), this step is also done at all shared edges between grids on the same resolution level in order to avoid roundoff errors, but it's not currently clear whether this will be necessary.)

<img width="855" alt="Screen Shot 2021-12-03 at 2 15 11 PM" src="https://user-images.githubusercontent.com/1781919/144539027-d413aabc-94a6-4e09-8d9c-a4f3539f9652.png">

**Comment by WeiqunZhang** [comment, marked as answer]
You can use
```
    void average_down_edges (const Array<const MultiFab*,AMREX_SPACEDIM>& fine,
                             const Array<MultiFab*,AMREX_SPACEDIM>& crse,
                             const IntVect& ratio,
                             int ngcrse = 0);
```
to average down fine edge data onto coarse.  Additionally, you can use
```
void MultiFab::OverrideSync (const Periodicity& period = Periodicity::NonPeriodic());
```
to make the data consistent on internal edges.

However, there is an issue with `average_down_edges` when there are periodic boundaries.  This can be fixed.  Will let you know.

**Reply by BenWibking** [reply]
That will make the EMFs consistent, but still leaves the magnetic fields inconsistent when doing AMR subcycling (since the difference in the curl of the EMF still has to be applied to the magnetic fields).

Some data structure is needed to store the difference in overlapping edges. I could allocate an edge-centered MultiFab on the coarse level for this, but that would waste a lot of memory, since only the edges at the coarse-fine interface actually need to be stored. Is there another data structure that I can repurpose for this? (Ideally it would be the same data structure used by YAFluxRegister, but maybe that only works for face-centered variables?)

**Reply by BenWibking** [reply]
I would think this also needs to be done in WarpX for non-electrostatic fields, but maybe there is no AMR subcycling in WarpX?

**Reply by WeiqunZhang** [reply]
OK. I misunderstood your question.  But if what you really care is curl E, not E itself, can you store curl E in YAFluxRegister?

**Reply by WeiqunZhang** [reply]
Never mind my message above.  I see the issue now.

**Reply by BenWibking** [reply]
Not the corner cases where there is an overlapping edge, but no overlapping face.

**Reply by WeiqunZhang** [reply]
Yes, I got it now.  We will need to write a new FluxRegister for this.

**Reply by BenWibking** [reply]
Ok, great. Maybe the Chombo implementation can be partially re-used?

**Reply by WeiqunZhang** [reply]
I just created an issue from this.

**Reply by BenWibking** [reply]
Ok, I've subscribed myself to the issue. If there's any input from me that would be helpful, let me know.
